### PR TITLE
FileWriter/Editor: fix bug in close with source

### LINKED ISSUE #236: FileWriter: confusing same message with `cancel` and `close` with save=0
both cases report same message with verbose level

### LINKED ISSUE #237: CAF::FileEditor from source broken
The diff from FileEditor with source is broken; regular close for FileEditor works by accident (it rereads the original contents). All thanks to typo.

### LINKED ISSUE #239: CAF::FileWriter: reported diff does not end up in verbose logfile
Could be an issue with `CAF::Reporter`

### PR DESCRIPTION
**Pull request by stdweird**
Fixes #237 
Also fixes #236 and fixes #239

### CODE REVIEW

**Comment by stdweird** [comment]
i could squash the `905` and `ee7` commits; the `ee7` one makes the problem clear and was what the old code did.

File: src/main/perl/FileWriter.pm [hunk 1]
```diff
@@ -215,21 +231,43 @@ sub close
...
+                $self->verbose("No original source content for $filename");
+            }
+        }
+
         # Always read and (try to) determine the diff
-        my $content_orig;
-        if (defined(*$self->{content_orig})) {
-            $self->debug(2, "Using original contents from $filename");
-            $content_orig = *$self->{content_orig};
+        my $original_content;
```
Comment (ned21) [inline src/main/perl/FileWriter.pm]: would it not be clearer to make this an elsif clause of the previous block and drop the ```!*$self->{original_from_source}``` test?

File: src/main/perl/FileWriter.pm [hunk 2]
```diff
@@ -215,21 +231,43 @@ sub close
...
+        if (defined(*$self->{original_content}) && !*$self->{original_from_source}) {
+            $self->debug(2, "Using existing original content for $filename");
+            $original_content = *$self->{original_content};
         } else {
             # missing_ok=1 mimics original LC::Check::file behaviour
-            $content_orig = $self->_read_contents($filename, event => \%event, missing_ok => 1);
+            $original_content = $self->_read_contents($filename, event => \%event, missing_ok => 1);
```
Comment (ned21) [inline src/main/perl/FileWriter.pm]: Why is this extra read needed here? The conditions under which this line is executed are not trivially fathomable (cf. comment above) so I think some extra comments here about how we can end up here and why we need this line would be useful.

File: src/main/perl/FileWriter.pm (hunk 1 above)
Comment (stdweird) [inline src/main/perl/FileWriter.pm]: hmmm, i'd rather not. the previous block is only about reporting, this block is about setting the content to use to determine change. i don't want to mix these, this is the critical bit of the close method.
maybe the attribute names can be changed?
i can also add that the first block is the "normal" path for FileEditor w/o source; the second path is the normal for FileWriter and FileEditor w source.
i just added the role and meaning of the attributes to the pod to make it clear what happens, i don't assume people should use them directly. maybe i should move the last block in the pod to regular comment?

File: src/main/perl/FileWriter.pm (hunk 2 above)
Comment (stdweird) [inline src/main/perl/FileWriter.pm]: it's not extra for regular FileWriter (this is the first and only time the original file is read) and regular FIleEditor (ie without source; that is the other block) use.

**Comment by stdweird** [comment]
@ned21 please rereview

**Comment by ned21** [comment]
Thanks for adding the extra comments to clarify that if block.

**Pull request merged**